File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        
        x = root

        while x:
            if p.val < x.val and q.val < x.val:
                x = x.left
            elif p.val > x.val and q.val > x.val:
                x = x.right
            else:
                return x

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:


    bool dfs(TreeNode* current, int& prev)
    {
        if(current == nullptr)
        {
            return true;
        }

        if(!dfs(current->left,prev))
        {
            return false;
        }

        if(current->val <= prev)
        {
            return false;
        }

        prev = current->val;

        return dfs(current->right, prev);
    }
    bool isValidBST(TreeNode* root) {

        int prev = INT_MIN;

        return dfs(root, prev);
        
    }
};
